mmengine/tune/tuner.py

```diff
@@ -0,0 +1,315 @@
+        from mmengine.runner import Runner
+
+        if is_main_process():
+            hparams_to_broadcast = [self._searcher.suggest()]
+        else:
+            hparams_to_broadcast = [None]  # type: ignore
+        broadcast_object_list(hparams_to_broadcast, src=0)
+        hparam = hparams_to_broadcast[0]
+
+        # Inject hyperparameters into runner config.
+        for k, v in hparam.items():
+            self.inject_config(self._runner_cfg, k, v)
+        runner = Runner.from_cfg(self._runner_cfg)
+        report_hook = ReportingHook(self._monitor, self._tuning_iter,
+                                    self._tuning_epoch, self._reporting_op)
+        runner.register_hook(report_hook, priority='VERY_LOW')
+        default_score = float('inf') if self._rule == 'less' else -float('inf')
+
+        # Run a trial.
+        # If an exception occurs during the trial, the score is set
+        # to default_score.
+        error = None
+        try:
+            runner.train()
+            score = report_hook.report_score()
+            if score is None or math.isnan(score) or math.isinf(score):
+                score = default_score
+            scores_to_broadcast = [score]
+        except Exception as e:
+            scores_to_broadcast = [default_score]
+            error = e
+
+        # Store the score between processes.
+        broadcast_object_list(scores_to_broadcast, src=0)
+        score = scores_to_broadcast[0]
+        if is_main_process():
+            self._searcher.record(hparam, score)
+        return hparam, score, error
+
+    def tune(self) -> Dict[str, Union[Dict[str, Any], float]]:
+        """Launch tuning.
+
+        Returns:
+        Dict[str, Union[Dict[str, Any], float]]:
+            A dictionary containing the best hyperparameters under the key
+            'hparam' and the corresponding score under the key 'score'.
+        """
+        self._logger.info(f'Starting tuning for {self._num_trials} trials...')
+        for trail_idx in range(self._num_trials):
+            hparam, score, error = self._run_trial()
+            log_msg = f'Trial [{trail_idx + 1}/{self._num_trials}]'
+            if error is not None:
+                log_msg += f' failed. Error: {error}'
+            else:
+                log_msg += f' finished. Score obtained: {score}'
+            log_msg += f' Hyperparameters used: {hparam}'
+            self._logger.info(log_msg)
+            self._history.append((hparam, score))
+
+        best_hparam: dict
+        best_score: float
+        if self._rule == 'greater':
+            best_hparam, best_score = max(self._history, key=lambda x: x[1])
+        else:
+            best_hparam, best_score = min(self._history, key=lambda x: x[1])
+        self._logger.info(f'Best hyperparameters obtained: {best_hparam}')
+        self._logger.info(f'Best score obtained: {best_score}')
+        self._logger.info('Tuning completed.')
+        return dict(hparam=best_hparam, score=best_score)
```

Comment:
In the Runner, several global variables are defined, including `Visualizer`, `Logger`, and `DefaultScope` (which inherits from ManagerMixin). We should clean these variables after each trial. Furthermore, it might be more advantageous to conduct the tuning process within a temporary directory and delete it once it is completed.

Reply:
Taking cues from your insights, I have incorporated a mechanism to clear global variables following each trial, referencing the [testing function](https://github.com/open-mmlab/mmengine/blob/c5274ba3268863aa17d15e0261f8c7f210dac59b/mmengine/testing/runner_test_case.py#L153) for the implementation. Additionally, I have configured the tuning process to transpire within a temporary directory, ensuring its deletion upon completion to maintain a clean workspace.